File: ZAJECIA_8/ZAD4/Edmonds-Karp.js
function edmondsKarp(graf, zrodlo, ujscie) {
    const n = graf.length;
    const grafRes = graf.map(row => row.slice());
    const rodzic = new Array(n);
    let maksPrzeplyw = 0;

    while (true) {
        rodzic.fill(-1);
        const kolejka = [zrodlo];
        while (kolejka.length > 0 && rodzic[ujscie] === -1) {
            const u = kolejka.shift();
            for (let v = 0; v < n; v++) {
                if (rodzic[v] === -1 && grafRes[u][v] > 0) {
                    rodzic[v] = u;
                    kolejka.push(v);
                }
            }
        }

        if (rodzic[ujscie] === -1) {
            break;
        }

        let przeplywSciezki = Infinity;
        for (let v = ujscie; v !== zrodlo; v = rodzic[v]) {
            const u = rodzic[v];
            przeplywSciezki = Math.min(przeplywSciezki, grafRes[u][v]);
        }
        for (let v = ujscie; v !== zrodlo; v = rodzic[v]) {
            const u = rodzic[v];
            grafRes[u][v] = grafRes[u][v] - przeplywSciezki;
            grafRes[v][u] = grafRes[v][u] + przeplywSciezki;
        }

        maksPrzeplyw = maksPrzeplyw + przeplywSciezki;
    }
    return maksPrzeplyw;
}

const graf =
    [
        // s  x   y   w  u  e  t
        [0, 20, 50, 0, 0, 0, 0],
        [0, 0, 0, 60, 50, 0, 0],
        [0, 10, 0, 30, 0, 0, 0],
        [0, 0, 0, 0, 10, 0, 20],
        [0, 0, 0, 0, 0, 80, 40],
        [0, 0, 0, 0, 0, 0, 40],
        [0, 0, 0, 0, 0, 0, 0]
    ];

const zrodlo = 0;
const ujscie = 6;
console.log("Maksymalny przeplyw:", edmondsKarp(graf, zrodlo, ujscie));

File: ZAJECIA_8/ZAD4/Edmonds-Karp.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

int edmonds_karp(vector<vector<int>> &graf, int zrodlo, int ujscie)
{
    int n = graf.size();
    vector<vector<int>> graf_residualny = graf;
    // Tablica rodzicow dla sciezki powiekszajacej
    vector<int> rodzic(n, -1);
    int maksymalnyPrzeplyw = 0;
    while (true)
    {
        queue<int> kolejka;
        kolejka.push(zrodlo);
        fill(rodzic.begin(), rodzic.end(), -1);
        while (!kolejka.empty() && rodzic[ujscie] == -1)
        {
            int u = kolejka.front();
            kolejka.pop();
            for (int v = 0; v < n; v++)
            {
                if (rodzic[v] == -1 && graf_residualny[u][v] > 0)
                {
                    rodzic[v] = u;
                    kolejka.push(v);
                }
            }
        }
        if (rodzic[ujscie] == -1)
        {
            break;
        }
        int przeplywSciezki = INT_MAX;
        for (int v = ujscie; v != zrodlo; v = rodzic[v])
        {
            int u = rodzic[v];
            przeplywSciezki = min(przeplywSciezki, graf_residualny[u][v]);
        }
        for (int v = ujscie; v != zrodlo; v = rodzic[v])
        {
            int u = rodzic[v];
            graf_residualny[u][v] = graf_residualny[u][v] - przeplywSciezki;
            graf_residualny[v][u] = graf_residualny[v][u] + przeplywSciezki;
        }
        maksymalnyPrzeplyw = maksymalnyPrzeplyw + przeplywSciezki;
    }
    return maksymalnyPrzeplyw;
}

int main()
{
    vector<vector<int>> graf =
        {
            // s  x   y   w  u  e  t
            {0, 20, 50, 0, 0, 0, 0},
            {0, 0, 0, 60, 50, 0, 0},
            {0, 10, 0, 30, 0, 0, 0},
            {0, 0, 0, 0, 10, 0, 20},
            {0, 0, 0, 0, 0, 80, 40},
            {0, 0, 0, 0, 0, 0, 40},
            {0, 0, 0, 0, 0, 0, 0}};
    int zrodlo = 0;
    int ujscie = 6;
    int maksymalnyPrzeplyw = edmonds_karp(graf, zrodlo, ujscie);
    cout << "Maksymalny przeplyw: " << maksymalnyPrzeplyw << endl;
    return 0;
}